### ISSUE TITLE
Clightgen with -normalize does not normalize structure assignments

### ISSUE THREAD

**Issue by MSoegtropIMC** [AbsInt/CompCert#299]
The Clightgen -normalize option does not work as expected in case of structure assignments. A simple example C program and the clightgen output created with

```
C:\bin\coq-8.9.1\bin\clightgen.exe -v -normalize -fstruct-passing -dclight Clightgen_struct_assign.c
```

are attached. I also included a variant of the function where I assign each field of the structure individually, which does the normalization as expected.

[Clightgen_struct_assign.zip](https://github.com/AbsInt/CompCert/files/3318352/Clightgen_struct_assign.zip)

This leads to issues with VST:

```
In nested Ltac calls to "start_function" and "check_normalized", last call failed.
Tactic failure: The expression  (_result)%expr
 contains internal memory dereferences, which suggests that you ran clightgen without the -normalize flag.  Print Info.normalized and make sure it has the value 'true' (level 97).
```

For your information I also attached a matching VST file.

This has been tested with the clightgen and VST bundled with Coq 8.9.1 for Windows.
Please note that the clightgen created file is slightly modified in the header to include the VST version of the compcert files.

**Comment by xavierleroy** [comment]
The repro case contains three C functions.  Could you clarify which function causes the error?

I'll go out on a limb and assume it is `complex_int_zero_ptr`, and more specifically the assignment `*_res  = result` at the end.  In Clight this is
```
    (Sassign
      (Ederef (Etempvar __res (tptr (Tstruct _complex_int_s noattr)))
        (Tstruct _complex_int_s noattr))
      (Evar _result (Tstruct _complex_int_s noattr)))))
```
I'm under the impression that VST's error message is misleading: the RHS of this assignment, `(Evar _result (Tstruct _complex_int_s noattr))`, does not "contain[s] internal memory dereferences", because variables of struct type evaluate to their addresses.  

At any rate I'm not sure what normalization should do differently to accommodate VST's constraints.  (Right now, only lvalue-to-rvalue conversions for scalar types are lifted out, because these are the only conversions that introduce memory loads.)

**Comment by MSoegtropIMC** [comment]
The cases `complex_int_zero` and `complex_int_zero_ptr` cannot be verified in VST cause of the issue. Clightgen seems to convert the structure return case into passing a pointer to the result structure, so the resulting AST is almost identical. The 3rd case `complex_int_zero_ptr_eva`, where I assign the field elements individually can be verified in VST.

Indeed I cannot tell if this is an issue in VST or in clightgen. But the difference in normalized AST code generated for the individual field assignment case and the full structure assignment case made me believe that the issue is with cligthgen. In the individual field assignment case normalization has the effect that each field element of the source structure is first copied into a temporary variable in a first statement and then this temporary is copied to the destination structure in a second statement. Below is the resulting code for one field:
```
      (Ssequence
        (Sset _t'2
          (Efield (Evar _result (Tstruct _complex_int_s noattr)) _real tint))
        (Sassign
          (Efield
            (Ederef (Etempvar __res (tptr (Tstruct _complex_int_s noattr)))
              (Tstruct _complex_int_s noattr)) _real tint)
          (Etempvar _t'2 tint)))
```
In the full structure assignment case, this copying to temporary variables is not done:
```
(Sassign
      (Ederef (Etempvar __res (tptr (Tstruct _complex_int_s noattr)))
        (Tstruct _complex_int_s noattr))
      (Evar _result (Tstruct _complex_int_s noattr)))
```
For VST the `(Evar _result (Tstruct _complex_int_s noattr)))` is illegal in the right hand side of an `Sassign`.

In my view the per value handling of structures should be similar to the handling of the individual fields, which is not the case.

@andrew-appel can you please comment in this?

**Comment by xavierleroy** [comment]
In the Clight semantics, assignment at composite types (structs and unions) has a specific semantics, which is similar to a memory copy, and cannot be expressed by assignment to individual fields.  (The latter is most obvious in case of unions.)  So, either VST has a specific reasoning rule for these kind of assignments, or you should not use them in VST-verified code.

**Comment by MSoegtropIMC** [comment]
I see. This explains the method clightgen uses to pass structures per value as arguments.

I guess VST would be happy if a similar method as for passing structures as arguments per value would be used for structure assignment. That is create a temporary array of unsigned integers (or other suitable types) large enough to cover the structure, then first copy the source structure to the temporary array and from there to the destination structure. But it most likely will be hard to verify such constructs with VST then, even if it would work. On the other hand one can likely automate this well.

Btw.: I am using per value passing of structures mostly for 2x16 bit ans 2x32 bit fixed point complex numbers. Most compilers implement this as passing 32 bit or 64 bit integers in a register, so this is very efficient. As you know I am using clightgen to verify code with VST which is not intended (at least for the time being) to be compiled with CompCert. With real values numerics this did work quite well, but for complex arithmetic the handling of structures in clightgen makes things a bit difficult.

**Comment by xavierleroy** [comment]
The semantics of structs/unions as first-class values is delicate.  See Robbert Krebber's CH2O (https://robbertkrebbers.nl/research/ch2o/) for a full formalization.  CompCert's semantics is much simpler, treating structs and unions like arrays, i.e. the value of a struct/union is a pointer to itself.  However it is not 100% conformant to ISO C.

The compilation of structs/unions as first-class values is equally painful, because the calling conventions vary greatly between platforms and ABIs.  Again, CompCert does some effort but is not 100% ABI conformant.

The bottom line is that in a verified library you should avoid passing structs/unions by value and pass them by reference  instead.  If your functions are `static inline`, inlining can remove some of the overhead of pass-by-reference.

**Comment by xavierleroy** [comment]
Closing this issue, as I consider that clightgen's behavior is consistent with CompCert's semantics for structs/unions as first-class values.

**Comment by MSoegtropIMC** [comment]
I need to discuss with Andrew how I can solve this. Entirely avoiding by value structures is an option, but not a nice one. The only area where I am using pass by value structures is complex arithmetic in C. It makes mathematical formulas much more readable if say `cadd` and `cmul` pass arguments and results by value. Otherwise all intermediate results need temporary variables. This would also make the separation logic of simple formulas quite a bit more complicated. Typically such functions are inline functions, so that from a performance point of view it doesn't make a difference.

**Comment by andrew-appel** [comment]
I'm sorry, in the last two weeks I haven't had a chance to comment on this thread.  Assuming we do not change the semantics of CompCert, then I see two other ways forward:

1.  Add a structure-assignment rule to VST, and prove it sound w.r.t. CompCert Clight operational semantics.  One challenge is that CompCert uses loadbytes/storebytes over the entire struct, whereas the "data_at" semantics we want to prove is field-by-field.   This challenge might be surmountable, especially for structs without any padding.   Then we'd need to add "Floyd" proof automation to handle the "forward" through such assignments.  

2.  Adjust `clightgen -normalize` to convert simple struct assignments into field-by-field copies.  It is already the case that the job of `-normalize` is to take Clight code that is already consistent with the C semantics, and make it more digestable.  Struct assignment is analogous:  the code that clightgen generates right now is, as Xavier writes, consistent with CompCert's semantics for structs/unions as first-class values; but it could be digested further for the convenience of verification.

Before I recommend whether to pursue option 1 or option 2, or some other option, I'll need time to study this, and I don't have time this week.
